akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala

```diff
@@ -36,240 +35,135 @@ private object PoolSlot {
-      if (canceled) onComplete()
 
-      context.become(unconnected)
-    }
+      // upstream pushes we create the inner stream if necessary or push if we're already connected
+      override def onPush(): Unit = {
+        def establishConnectionFlow() = {
+          connectionFlowSource = new SubSourceOutlet[HttpRequest]("RequestSource")
+          connectionFlowSource.setHandler(connectionOutFlowHandler)
 
-    override def onComplete(): Unit = {
-      exposedPublisher.shutdown(None)
-      super.onComplete()
-      shutdown()
-    }
+          connectionFlowSink = new SubSinkInlet[HttpResponse]("ResponseSink")
+          connectionFlowSink.setHandler(connectionInFlowHandler)
 
-    override def onError(cause: Throwable): Unit = {
-      exposedPublisher.shutdown(Some(cause))
-      super.onError(cause)
-      shutdown()
-    }
+          isConnected = true
 
-    def shutdown(): Unit = context.stop(self)
-  }
+          Source.fromGraph(connectionFlowSource.source)
+            .via(connectionFlow).runWith(Sink.fromGraph(connectionFlowSink.sink))(subFusingMaterializer)
 
-  private case class FromConnection(ev: Any) extends NoSerializationVerificationNeeded
-
-  private class FlowInportActor(slotProcessor: ActorRef) extends ActorPublisher[HttpRequest] with ActorLogging {
-    def receive: Receive = {
-      case ev: Request            ⇒ slotProcessor ! FromConnection(ev)
-      case OnNext(r: HttpRequest) ⇒ onNext(r)
-      case OnComplete             ⇒ onCompleteThenStop()
-      case OnError(e)             ⇒ onErrorThenStop(e)
-      case Cancel ⇒
-        slotProcessor ! FromConnection(Cancel)
-        context.stop(self)
-    }
-  }
-  def flowInportProps(s: ActorRef) = Props(new FlowInportActor(s)).withDeploy(Deploy.local)
-
-  private class FlowOutportActor(slotProcessor: ActorRef) extends ActorSubscriber with ActorLogging {
-    def requestStrategy = ZeroRequestStrategy
-    def receive: Receive = {
-      case Request(n) ⇒ request(n)
-      case Cancel     ⇒ cancel()
-      case ev: OnNext ⇒ slotProcessor ! FromConnection(ev)
-      case ev @ (OnComplete | OnError(_)) ⇒
-        slotProcessor ! FromConnection(ev)
-        context.stop(self)
+          connectionFlowSink.pull()
+        }
+
+        grab(in) match {
+          case ConnectEagerlyCommand ⇒
+            if (!isConnected) establishConnectionFlow()
+
+            emit(eventsOut, ConnectedEagerly(slotIx))
+
+          case DispatchCommand(rc: RequestContext) ⇒
+            if (isConnected) {
+              inflightRequests.add(rc)
+              connectionFlowSource.push(rc.request)
+            } else {
+              firstRequest = rc
+              establishConnectionFlow()
+            }
+        }
+      }
+
+      setHandler(in, this)
+
+      setHandler(responsesOut, new OutHandler {
+        override def onPull(): Unit = {
+          // downstream pulls, if connected we pull inner
+          if (isConnected) connectionFlowSink.pull()
```

Comment:
the `isConnected` case doesn't have this problem since we have a simple 1-1, the guard is on the disconnected because the `eventsOut` doesn't always get an element.